### DISCUSSION TITLE
Why NEAR does not expose iterators

### DISCUSSION THREAD

**Discussion by maxzaver** [near/NEPs#160] [General]
As you might have noticed, NEAR does not allow Wasm contracts iterate over the storage anymore: https://github.com/near/nearcore/blob/1e88c6c699e3a2c49c6cbd09d284f69885982e80/runtime/near-vm-logic/src/logic.rs#L2226-L2298

There are several reasons:
* Contracts is the only non-upgradable component in our blockchain (except some very special-cased contracts like staking pool). If we introduce some host function we cannot remove it later without risking to break a large number of contracts in our blockchain. We don't want to introduce iterators, because we might want to switch to different state designs in the future that do not support iterators or have different iterator behavior (e.g. in Merkle Patricia tree keys with the same prefix belong to the same subtree). We are discussing Urkel, AVL, and variations of binary trees. Future modifications of the state storage might allow us to unlock higher TPS and so we want to reserve the possibility for state design iterations.
* We might enable challenges and slashing in the future. Challenges require state witnesses which are difficult to construct if contracts have access to iterators. We want to avoid this complexity.

Therefore, from the contract perspective there should be only one way to access the state: key-value lookup/read/write, without any advanced operations that are usually available in key-value storage, e.g. iterators.

**Comment by artob** [comment]
A follow-up from @ilblackdragon: https://github.com/near/NEPs/discussions/159

**Comment by ilblackdragon** [comment]
@nearmax Thanks for your response.
How about we move this post to protocol discussions in NEPs (protocol repo)? Also if there are other protocol related topics - let's also move them there.

**Reply by bowenwang1996** [reply]
@ilblackdragon done
